ratelimit/backends/simple.py

```diff
@@ -0,0 +1,107 @@
+import asyncio
+from collections import defaultdict
+from threading import Lock
+from typing import Dict, List, Optional
+
+from ..rule import Rule
+from . import BaseBackend
+
+
+def synchronized(func):
+    def wrapper(*args, **kwargs):
+        lock = Lock()
+        try:
+            lock.acquire()
+            return func(*args, **kwargs)
+        finally:
+            lock.release()
+
+    return wrapper
+
+
+class MemoryBackend(BaseBackend):
+    """simple limiter with memory"""
+
+    def __init__(self):
+        # user: deadline
+        self.blocked_users: Dict[str, int] = {}
+        # path: {rule_key: (limit, timestamp)}
+        self.blocks: Dict[str, Dict[str, List[int]]] = defaultdict(dict)
```

Comment:
Maybe I didn't express it clearly, I mean that each pair of blocked_users and blocks corresponds to a `Lock` 😂 We should lock the multi-threaded concurrent operations on the data structure, right?

Reply:
My mistake. But the `dict` is thread-safe, the `lock` is unnecessary maybe ?

Reply:
> My mistake. But the dict is thread-safe, the lock is unnecessary maybe ?

`self.blocks[path].pop(rule_key, None)` is not thread-safe. 

---

`retry_after` need a Lock to keep number decr is thread-safe